homework/zadanie5_false_sharing.cpp

```diff
@@ -0,0 +1,158 @@
+#include <iostream>
+#include <thread>
+#include <chrono>
+#include <new>
+#include <vector>
+
+struct alignas(64) Int64{
+    int a;
+};
+
+struct Int{
+    int a;
+};
+
+template<typename T>
+void fill(T* array, int begin, int end, int value){
+    for (int i = begin; i < end; i++){
+        array->a = value;
+        array++;
+    }
+}
+
+int main(){
+
+    int size = 100000;
+    int small_size = 1000;
+    int fill_it_up = 2;
+
+    Int array[size];
+    Int64 array64[size];
+
+    Int small_array[small_size];
+    Int64 small_array64[small_size];
+    
+    std::cout<< "\nSize of array:   " << sizeof(array) << "\n";
+    std::cout<< "Size of array64: " << sizeof(array64) << "\n";
+    std::cout<< "\nSize of small_array:   " << sizeof(small_array) << "\n";
+    std::cout<< "Size of small_array64: " << sizeof(small_array64) << "\n";
+
+    std::cout<< "\n---------------------------------------default size: \n\n";
+
+    auto start = std::chrono::high_resolution_clock::now();
+    fill(array, 0, size, fill_it_up);
+    auto stop = std::chrono::high_resolution_clock::now();
+    std::cout<< "Time of function -> fill(array_without_padding): " << std::chrono::duration_cast<std::chrono::microseconds>(stop - start).count() << "\n\n";
+
+    start = std::chrono::high_resolution_clock::now();
+    fill(array64, 0, size, fill_it_up);
+    stop = std::chrono::high_resolution_clock::now();
+    std::cout<< "Time of function -> fill(array_with_padding): " << std::chrono::duration_cast<std::chrono::microseconds>(stop - start).count() << "\n\n";
+
+    start = std::chrono::high_resolution_clock::now();
+    std::thread t([&](){fill(array, 0, size-1, fill_it_up);});
+    std::thread t1([&](){fill(array, size/2, size, fill_it_up);});
```

Comment:
przy takim podziale zadań właśnie nie zaobserwujesz false sharingu. To też widać na wynikach, że wersja wielowątkowa dla dużych rozmiarów i tak jest szybsza od zwykłej. No i w przypadku paddingu nie uzyskujesz przyspieszenia.
Musiałbyś zadania podzielić tak, że 1 wątek dostaje np. parzyste indeksy tablisy do wypełnienie, a drugi nieparzyste. Wtedy piszą one po komórkach pamięci leżącej obok siebie, więc będą sobie unieważniać zawartość pamięci cache, a więc będzie spowolnienie. Spróbuj i zobacz :)

Reply:
Poprawione. Dokladnie mialo to miejsce we wczesniejszej wersji programu, a w obecnej widzimy ze uruchomienie zadania na watkach skutkuje zwiekszonym czasem dzialania,